Support for shadow offset in box style

pouleyKetchoupp opened the pull request:
This change adds an offset property to box style shadows that can be used to adjust the position of the shadow behind the panel.

<img src="https://user-images.githubusercontent.com/1075032/53761501-fddb5500-3ec5-11e9-91ae-14592ee6c954.png" width=300 />

<img src="https://user-images.githubusercontent.com/1075032/53761511-00d64580-3ec6-11e9-8bc8-5ae0d850e0cc.png" width=300 />

pouleyKetchoupp committed:
Support for shadow offset in box style

toger5 commented:
Looks good.
Maybe it still would be better, to also give the shadow a fill. So that the offset can be greater than the shadow width.
Than there are less condintions which is always a little annoying if u want to set something to a high value, but it doesnt work.
![](https://htmldog.com/figures/boxshadow.png)
A shadow offset which looks like the one in the picture schould be possible i think.
What do u think?

pouleyKetchoupp commented:
@toger5 Yeah that makes sense. So far I've only modified the existing implementation based on a "ring" shadow to make it simpler, but instead I could draw a full shadow to handle offsets larger than the width better.

It would look like that with Transparent/Opaque box fill:
<img src="https://user-images.githubusercontent.com/1075032/53791768-770c9380-3f2a-11e9-9c1e-9e420dbc5a24.png" width = 400/>

So it will slightly change the current behavior, as currently you can't see the shadow through a transparent box, but I think it's for the best as it's more likely how you would expect a shadow to work.

Seems good to you?

toger5 commented:
I think you didn't force pull your changes. Or did u open a new pr?

By the way why is the pr from nekomatata but you did the commit. (of course that is fine. but I don't really understand why?) maybe something got messed up there and thats why the pr is still using the ring version you had before.

pouleyKetchoupp commented:
Sorry for the confusion, I was just waiting for validation before actually making the change, because of the slightly different behavior in case of transparency.

I guess it's a yes, so I'll go for it in this PR :)

Also, I'm using the fork from Nekomatata because it's the common one for our little studio, even though I'm the only one submitting changes for now.

toger5 commented:
Yea the bahaviour (also with transparent sb's) is great like you posted.
You can just force push so i can look over the code.

pouleyKetchoupp commented:
I've just pushed this change to draw the full shadow.

Also, I've fixed two glitches around transparent panels at the same time:

**glitch 1:**
Panel with transparency and rounded corners (more visible with less corner details)
<img src="https://user-images.githubusercontent.com/1075032/53961031-f8f5eb80-40e7-11e9-9b6a-de2f49122817.png" width=300 />

fixed with
`index < added_vert_count / 2`
instead of
`index <= added_vert_count / 2`
when adding indices to fill the center - line 644 (it was adding extra faces)

**glitch 2:**
Panel with transparency and rounded corners has a thin line of different color around the border
<img src="https://user-images.githubusercontent.com/1075032/53961146-3bb7c380-40e8-11e9-97e6-0c1e187675ef.png" width=300 />

fixed with
```
	for (int i = 0; i < vert_count; i++) {
		indices.push_back(vert_offset + ((i + 0) % vert_count));
		indices.push_back(vert_offset + ((i + 2) % vert_count));
		indices.push_back(vert_offset + ((i + 1) % vert_count));
	}
```
instead of
```
	for (int i = 0; i < vert_count; i++) {
		//poly 1
		indices.push_back(vert_offset + ((i + 0) % vert_count));
		indices.push_back(vert_offset + ((i + 2) % vert_count));
		indices.push_back(vert_offset + ((i + 1) % vert_count));
		//poly 2
		indices.push_back(vert_offset + ((i + 1) % vert_count));
		indices.push_back(vert_offset + ((i + 2) % vert_count));
		indices.push_back(vert_offset + ((i + 3) % vert_count));
	}
```
when adding indices for the ring itself - line 637 (it was doubling all the faces)

pouleyKetchoupp commented:
And I've spotted a different glitch, this one around shadows. I haven't found a proper fix for it though.

**glitch 3:**
Panel with no rounded corner, shadow is not rectangle shape
<img src="https://user-images.githubusercontent.com/1075032/53961427-dca67e80-40e8-11e9-8d1c-0d907e911ea4.png" width =300 />

Would you like me to open an issue for this one?

toger5 commented:
@pouleyKetchoupp thats some really good stuff you found:

Glitch3: that is actually intentional. I wanted the shadow to be rounded. But i think that should be an option.

Glitch2: so the naming i choose is really poor than, instead of vert cpunt it should be tri_count. Can u try it, since i am not sure if every tri is actually drawn twice. I am looping through the rects and for each rect i draw teo triangles/polys. The line might be a result from the border

pouleyKetchoupp commented:
@toger5 Thanks for your comments!

Glitch3: Ok, that makes sense. No need to change this behavior for now then, and if someone needs the option, it can be implemented as a separate PR.

Glitch2: I don't think the line is due to borders, all the border widths are set to 0 in the screenshots I've posted before. And here's how it looks with the same settings after my change:
<img src="https://user-images.githubusercontent.com/1075032/54489283-e8ddc900-48aa-11e9-9eaf-324651ebc997.png" width=300 />
So it seems everything's drawn properly, and looks very similar to the way things are drawn with no corner radius (after and before my change):
<img src="https://user-images.githubusercontent.com/1075032/54489365-bd0f1300-48ab-11e9-8fcc-5ac847c17b9b.png" width=300 />
Unless I'm missing something, it really looks like the second half of the loop was mostly re-drawing the same triangles (since vertex indices are incremented by 1 on each iteration).

toger5 commented:
@pouleyKetchoupp okay, now I finally found some time to investigate. When drawing the ring (with polygons):
```
for (int i = 0; i < vert_count; i++) {
		//poly 1
		indices.push_back(vert_offset + ((i + 0) % vert_count));
		indices.push_back(vert_offset + ((i + 2) % vert_count));
		indices.push_back(vert_offset + ((i + 1) % vert_count));
		//poly 2
		indices.push_back(vert_offset + ((i + 1) % vert_count));
		indices.push_back(vert_offset + ((i + 2) % vert_count));
		indices.push_back(vert_offset + ((i + 3) % vert_count));
	}
```
we have to add poly 1 and 2.
<img width="366" alt="Screen Shot 2019-03-17 at 18 51 19" src="https://user-images.githubusercontent.com/16718859/54495640-c7e69980-48e5-11e9-9043-701499eacf5c.png">
Poly one is the red one and poly two the blue one. the vert_count property holds the number of vertices in the outer OR the inner part of the ring. (In the case of the illustration `vert_count = 6`)

The light line around it is caused by the anti aliasing:
<img width="300" alt="Screen Shot 2019-03-17 at 18 54 36" src="https://user-images.githubusercontent.com/16718859/54495663-1f850500-48e6-11e9-911d-0980e0f19b10.png">
The screenshot shows the blending from the inner part to the outer one. drawing that blend with only one pixel width results in smoother edges especially when having rounded corners.
My guess is, that I made the mistake to not blend from the correct color, instead I use the color without alpha. so the aa results in a bright 1px line around the sb.
Without AntiAliasing:
<img width="300" alt="Screen Shot 2019-03-17 at 18 58 22" src="https://user-images.githubusercontent.com/16718859/54495715-a934d280-48e6-11e9-9260-c13cfde00db5.png">

Can u check how it looks when activating a border width of (lets say) 10, and only drawing one poly per  vert?

And maybe you find a fix for the aa (so it does use the correct vert color for transparent bgcolors)?

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -618,17 +630,28 @@ inline void draw_ring(Vector<Vector2> &verts, Vector<int> &indices, Vector<Color
-		indices.push_back(vert_offset + ((i + 3) % vert_count));
+	}
+
+	if (fill_center) {
```
Okay now I understand, that you put the fill center inside the draw ring func.

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -709,23 +746,9 @@ void StyleBoxFlat::draw(RID p_canvas_item, const Rect2 &p_rect) const {
+				infill_rect, no_border, &bg_color, &bg_color, corner_detail, true);
 	}
 
 	if (aa_on) {
```
this is the actual issue of the one pixel light border. 
`Color c = border_color.read().ptr()[I];` The color is based on the border color. which is one correct if `border_width > 0`

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -618,17 +630,28 @@ inline void draw_ring(Vector<Vector2> &verts, Vector<int> &indices, Vector<Color
 	}
 
-	int vert_count = (adapted_corner_detail + 1) * 4 * 2;
+	int vert_count = verts.size() - vert_offset;
```
maybe rename to `ring_vert_count`

toger5 commented on scene/resources/style_box.cpp:
```diff
@@ -618,17 +630,28 @@ inline void draw_ring(Vector<Vector2> &verts, Vector<int> &indices, Vector<Color
 	//fill the indices and the colors for the border
 	for (int i = 0; i < vert_count; i++) {
-		//poly 1
 		indices.push_back(vert_offset + ((i + 0) % vert_count));
```
here you are totally right, that everything gets drawn twice. I would have needed to do:
`int vert_count = (adapted_corner_detail + 1) * 4;` without the `* 2`.

As I explained in the post before, the issue with the bright line is still caused by the Anti aliasing. This effect causes a bug, when using the same (semi transparent) color for the border and the infill. Than the border is drawn twice -> is darker.

toger5 replied:
Here is a screenshot:
<img width="327" alt="Screen Shot 2019-03-17 at 19 16 08" src="https://user-images.githubusercontent.com/16718859/54495947-8f48bf00-48e9-11e9-8e36-53bbc1eb25e9.png">

pouleyKetchoupp committed:
Properly handle different border widths in rounded corner aa

Note: removed (unused and not implemented) support for multiple border colors

pouleyKetchoupp commented:
@toger5 I've just pushed a new version that fixes the anti-aliasing when using different widths for the border. Thanks for all the details, that was very helpful.

Here's how it looks now with aa on and various border settings:
<img src="https://user-images.githubusercontent.com/1075032/54598336-9205fa00-4a38-11e9-9d85-bb1c8a5543f0.png"  width=300 />
<img src="https://user-images.githubusercontent.com/1075032/54598340-93cfbd80-4a38-11e9-8f15-c64dc60f661b.png"  width=300 />

On a side note, I've removed the support for multiple colors for borders in StyleBoxFlat, for simplification purpose. The reason is it was used nowhere, not accessible by script, and the ring code actually didn't handle it (it was setting the first color to every vertex).

pouleyKetchoupp commented:
I've pushed the aa changes as a separate commit to make it easier to check differences with the actual shadow offset changes, but I can squash them together later if needed.

pouleyKetchoupp commented on scene/resources/style_box.cpp:
```diff
@@ -552,7 +551,7 @@ inline void set_inner_corner_radius(const Rect2 style_rect, const Rect2 inner_re
 
 	//tr
-	rad = MIN(border_top, border_bottom);
+	rad = MIN(border_top, border_right);
```
This is a little fix for another glitch that could cause the top right rounded corner to be drawn with the wrong size in some cases.

toger5 replied:
ups that was obviously a mistake.

pouleyKetchoupp committed:
Updated StyleBoxFlat documentation

reduz commented:
looks great, as long as @toger5 is happy

akien-mga commented:
Changes seem fine to me too. @toger5 you can merge when it's ready.
I'll leave it to you two to see if the commits should be squashed or if they're fine as 3 separate commits.

toger5 commented:
Really like the changes/fixes.
the commit history makes sense since the second is a fix and the last separates the doc.
